// rsuite/esm/FlexboxGrid/FlexboxGridItem.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import PropTypes from 'prop-types';
import { useClassNames } from '../utils';
var defaultProps = {
  as: 'div',
  classPrefix: 'flex-box-grid-item',
  colspan: 0,
  order: 0
};
var FlexboxGridItem = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _withClassPrefix;

  var Component = props.as,
      className = props.className,
      classPrefix = props.classPrefix,
      colspan = props.colspan,
      order = props.order,
      rest = _objectWithoutPropertiesLoose(props, ["as", "className", "classPrefix", "colspan", "order"]);

  var _useClassNames = useClassNames(classPrefix),
      merge = _useClassNames.merge,
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = merge(className, withClassPrefix((_withClassPrefix = {}, _withClassPrefix[colspan] = colspan >= 0, _withClassPrefix["order-" + order] = order, _withClassPrefix)));
  return /*#__PURE__*/React.createElement(Component, _extends({
    ref: ref
  }, rest, {
    className: classes
  }));
});
FlexboxGridItem.displayName = 'FlexboxGridItem';
FlexboxGridItem.defaultProps = defaultProps;
FlexboxGridItem.propTypes = {
  as: PropTypes.elementType,
  className: PropTypes.string,
  colspan: PropTypes.number,
  order: PropTypes.number,
  classPrefix: PropTypes.string
};
export default FlexboxGridItem;
