// rsuite/esm/FlexboxGrid/FlexboxGrid.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import PropTypes from 'prop-types';
import { useClassNames } from '../utils';
import FlexboxGridItem from './FlexboxGridItem';
var defaultProps = {
  as: 'div',
  classPrefix: 'flex-box-grid',
  align: 'top',
  justify: 'start'
};
var FlexboxGrid = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Component = props.as,
      className = props.className,
      classPrefix = props.classPrefix,
      align = props.align,
      justify = props.justify,
      rest = _objectWithoutPropertiesLoose(props, ["as", "className", "classPrefix", "align", "justify"]);

  var _useClassNames = useClassNames(classPrefix),
      merge = _useClassNames.merge,
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = merge(className, withClassPrefix(align, justify));
  return /*#__PURE__*/React.createElement(Component, _extends({}, rest, {
    ref: ref,
    className: classes
  }));
});
FlexboxGrid.Item = FlexboxGridItem;
FlexboxGrid.displayName = 'FlexboxGrid';
FlexboxGrid.defaultProps = defaultProps;
FlexboxGrid.propTypes = {
  className: PropTypes.string,
  classPrefix: PropTypes.string,
  align: PropTypes.oneOf(['top', 'middle', 'bottom']),
  justify: PropTypes.oneOf(['start', 'end', 'center', 'space-around', 'space-between'])
};
export default FlexboxGrid;
