// rsuite/esm/Placeholder/PlaceholderParagraph.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { useClassNames } from '../utils';
var defaultProps = {
  as: 'div',
  classPrefix: 'placeholder',
  rows: 2,
  rowHeight: 10,
  rowMargin: 20
};
var PlaceholderParagraph = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Component = props.as,
      className = props.className,
      rows = props.rows,
      rowHeight = props.rowHeight,
      rowMargin = props.rowMargin,
      graph = props.graph,
      active = props.active,
      classPrefix = props.classPrefix,
      rest = _objectWithoutPropertiesLoose(props, ["as", "className", "rows", "rowHeight", "rowMargin", "graph", "active", "classPrefix"]);

  var _useClassNames = useClassNames(classPrefix),
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix,
      withClassPrefix = _useClassNames.withClassPrefix;

  var graphShape = graph === true ? 'square' : graph;
  var rowElements = useMemo(function () {
    var rowArr = [];

    for (var i = 0; i < rows; i++) {
      var styles = {
        width: Math.random() * 75 + 25 + "%",
        height: rowHeight,
        marginTop: i > 0 ? rowMargin : Number(rowMargin) / 2
      };
      rowArr.push( /*#__PURE__*/React.createElement("p", {
        key: i,
        style: styles
      }));
    }

    return rowArr;
  }, [rowHeight, rowMargin, rows]);
  var classes = merge(className, withClassPrefix('paragraph', {
    active: active
  }));
  var graphClasses = prefix('paragraph-graph', "paragraph-graph-" + graphShape);
  return /*#__PURE__*/React.createElement(Component, _extends({}, rest, {
    ref: ref,
    className: classes
  }), graphShape && /*#__PURE__*/React.createElement("div", {
    className: graphClasses
  }, /*#__PURE__*/React.createElement("span", {
    className: prefix('paragraph-graph-inner')
  })), /*#__PURE__*/React.createElement("div", {
    className: prefix('paragraph-rows')
  }, rowElements));
});
PlaceholderParagraph.displayName = 'PlaceholderParagraph';
PlaceholderParagraph.defaultProps = defaultProps;
PlaceholderParagraph.propTypes = {
  className: PropTypes.string,
  classPrefix: PropTypes.string,
  rows: PropTypes.number,
  rowHeight: PropTypes.number,
  rowMargin: PropTypes.number,
  graph: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['circle', 'square', 'image'])]),
  active: PropTypes.bool
};
export default PlaceholderParagraph;
