// rsuite/esm/Placeholder/PlaceholderGrid.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useClassNames } from '../utils';
var defaultProps = {
  as: 'div',
  classPrefix: 'placeholder',
  rows: 5,
  columns: 5,
  rowHeight: 10,
  rowMargin: 20
};
var PlaceholderGrid = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Component = props.as,
      className = props.className,
      classPrefix = props.classPrefix,
      rows = props.rows,
      columns = props.columns,
      rowHeight = props.rowHeight,
      rowMargin = props.rowMargin,
      active = props.active,
      rest = _objectWithoutPropertiesLoose(props, ["as", "className", "classPrefix", "rows", "columns", "rowHeight", "rowMargin", "active"]);

  var _useClassNames = useClassNames(classPrefix),
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix,
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = merge(className, withClassPrefix('grid', {
    active: active
  }));
  var colItems = [];
  var firstRowItemWidth = Math.random() * 30 + 30;
  var itemWidth = firstRowItemWidth / 2;

  for (var i = 0; i < columns; i++) {
    var rowItems = [];

    for (var j = 0; j < rows; j++) {
      var widthPercent = Math.random() * 50 + 10; // when first column

      if (i > 0) {
        // when other columns
        widthPercent = j > 0 ? itemWidth : firstRowItemWidth;
      }

      rowItems.push( /*#__PURE__*/React.createElement("p", {
        key: j,
        style: {
          width: widthPercent + "%",
          height: rowHeight,
          marginTop: j > 0 ? rowMargin : null
        }
      }));
    }

    colItems.push( /*#__PURE__*/React.createElement("div", {
      key: i,
      className: classNames(prefix('grid-col'))
    }, rowItems));
  }

  return /*#__PURE__*/React.createElement(Component, _extends({}, rest, {
    ref: ref,
    className: classes
  }), colItems);
});
PlaceholderGrid.displayName = 'PlaceholderGrid';
PlaceholderGrid.defaultProps = defaultProps;
PlaceholderGrid.propTypes = {
  className: PropTypes.string,
  classPrefix: PropTypes.string,
  rows: PropTypes.number,
  columns: PropTypes.number,
  rowHeight: PropTypes.number,
  rowMargin: PropTypes.number,
  active: PropTypes.bool
};
export default PlaceholderGrid;
