// rsuite/cjs/Slider/Mark.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

var defaultProps = {
  as: 'span',
  classPrefix: 'slider-mark'
};

var Mark = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var Component = props.as,
      mark = props.mark,
      last = props.last,
      classPrefix = props.classPrefix,
      className = props.className,
      renderMark = props.renderMark;

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix,
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = merge(className, withClassPrefix({
    last: last
  }));

  if (renderMark) {
    return /*#__PURE__*/_react.default.createElement(Component, {
      ref: ref,
      className: classes
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: prefix('content')
    }, renderMark(mark)));
  }

  return null;
});

Mark.displayName = 'Mark';
Mark.defaultProps = defaultProps;
Mark.propTypes = {
  as: _propTypes.default.elementType,
  classPrefix: _propTypes.default.string,
  className: _propTypes.default.string,
  mark: _propTypes.default.number,
  last: _propTypes.default.bool,
  renderMark: _propTypes.default.func
};
var _default = Mark;
exports.default = _default;
