// rsuite/cjs/Slider/ProgressBar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

var defaultProps = {
  as: 'div',
  classPrefix: 'slider-progress-bar'
};

var ProgressBar = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _extends2;

  var Component = props.as,
      classPrefix = props.classPrefix,
      vertical = props.vertical,
      rtl = props.rtl,
      _props$end = props.end,
      end = _props$end === void 0 ? 0 : _props$end,
      _props$start = props.start,
      start = _props$start === void 0 ? 0 : _props$start,
      style = props.style,
      className = props.className;

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      merge = _useClassNames.merge,
      withClassPrefix = _useClassNames.withClassPrefix;

  var sizeKey = vertical ? 'height' : 'width';
  var dirKey = rtl ? 'right' : 'left';
  var startKey = vertical ? 'top' : dirKey;
  var styles = (0, _extends3.default)({}, style, (_extends2 = {}, _extends2[startKey] = start + "%", _extends2[sizeKey] = end - start + "%", _extends2));
  var classes = merge(className, withClassPrefix());
  return /*#__PURE__*/_react.default.createElement(Component, {
    ref: ref,
    style: styles,
    className: classes
  });
});

ProgressBar.displayName = 'ProgressBar';
ProgressBar.defaultProps = defaultProps;
ProgressBar.propTypes = {
  as: _propTypes.default.elementType,
  classPrefix: _propTypes.default.string,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  vertical: _propTypes.default.bool,
  rtl: _propTypes.default.bool,
  start: _propTypes.default.number,
  end: _propTypes.default.number
};
var _default = ProgressBar;
exports.default = _default;
