// rsuite/esm/utils/createComponent.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import kebabCase from 'lodash/kebabCase';
import PropTypes from 'prop-types';
import { useClassNames } from '../utils';

/**
 * Create a component with `classPrefix` and `as` attributes.
 * @param defaultProps
 */
function createComponent(defaultProps) {
  var name = defaultProps.name,
      componentAs = defaultProps.componentAs,
      componentClassPrefix = defaultProps.componentClassPrefix,
      componentRestProps = _objectWithoutPropertiesLoose(defaultProps, ["name", "componentAs", "componentClassPrefix"]);

  var Component = /*#__PURE__*/React.forwardRef(function (props, ref) {
    var Component = props.as,
        classPrefix = props.classPrefix,
        className = props.className,
        role = props.role,
        rest = _objectWithoutPropertiesLoose(props, ["as", "classPrefix", "className", "role"]);

    var _useClassNames = useClassNames(classPrefix),
        withClassPrefix = _useClassNames.withClassPrefix,
        merge = _useClassNames.merge;

    var classes = merge(className, withClassPrefix());
    return /*#__PURE__*/React.createElement(Component, _extends({}, rest, {
      role: role,
      ref: ref,
      className: classes
    }));
  });
  Component.displayName = name;
  Component.propTypes = {
    as: PropTypes.elementType,
    className: PropTypes.string,
    classPrefix: PropTypes.string,
    children: PropTypes.node
  };
  Component.defaultProps = _extends({}, componentRestProps, {
    as: componentAs || 'div',
    classPrefix: componentClassPrefix || kebabCase(name)
  });
  return Component;
}

export default createComponent;
