// rsuite/esm/Calendar/Table.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import PropTypes from 'prop-types';
import TableRow from './TableRow';
import TableHeaderRow from './TableHeaderRow';
import { useClassNames } from '../utils';
var defaultProps = {
  as: 'div',
  classPrefix: 'calendar-table',
  rows: []
};
var Table = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Component = props.as,
      className = props.className,
      classPrefix = props.classPrefix,
      rows = props.rows,
      rest = _objectWithoutPropertiesLoose(props, ["as", "className", "classPrefix", "rows"]);

  var _useClassNames = useClassNames(classPrefix),
      merge = _useClassNames.merge,
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = merge(className, withClassPrefix());
  return /*#__PURE__*/React.createElement(Component, _extends({
    role: "table"
  }, rest, {
    ref: ref,
    className: classes
  }), /*#__PURE__*/React.createElement(TableHeaderRow, null), rows.map(function (week, index) {
    return /*#__PURE__*/React.createElement(TableRow, {
      key: index,
      weekendDate: week
    });
  }));
});
Table.displayName = 'Table';
Table.defaultProps = defaultProps;
Table.propTypes = {
  rows: PropTypes.array,
  className: PropTypes.string,
  classPrefix: PropTypes.string
};
export default Table;
