// rsuite/cjs/Placeholder/PlaceholderGrid.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../utils");

var defaultProps = {
  as: 'div',
  classPrefix: 'placeholder',
  rows: 5,
  columns: 5,
  rowHeight: 10,
  rowMargin: 20
};

var PlaceholderGrid = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var Component = props.as,
      className = props.className,
      classPrefix = props.classPrefix,
      rows = props.rows,
      columns = props.columns,
      rowHeight = props.rowHeight,
      rowMargin = props.rowMargin,
      active = props.active,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["as", "className", "classPrefix", "rows", "columns", "rowHeight", "rowMargin", "active"]);

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      merge = _useClassNames.merge,
      prefix = _useClassNames.prefix,
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = merge(className, withClassPrefix('grid', {
    active: active
  }));
  var colItems = [];
  var firstRowItemWidth = Math.random() * 30 + 30;
  var itemWidth = firstRowItemWidth / 2;

  for (var i = 0; i < columns; i++) {
    var rowItems = [];

    for (var j = 0; j < rows; j++) {
      var widthPercent = Math.random() * 50 + 10; // when first column

      if (i > 0) {
        // when other columns
        widthPercent = j > 0 ? itemWidth : firstRowItemWidth;
      }

      rowItems.push( /*#__PURE__*/_react.default.createElement("p", {
        key: j,
        style: {
          width: widthPercent + "%",
          height: rowHeight,
          marginTop: j > 0 ? rowMargin : null
        }
      }));
    }

    colItems.push( /*#__PURE__*/_react.default.createElement("div", {
      key: i,
      className: (0, _classnames.default)(prefix('grid-col'))
    }, rowItems));
  }

  return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({}, rest, {
    ref: ref,
    className: classes
  }), colItems);
});

PlaceholderGrid.displayName = 'PlaceholderGrid';
PlaceholderGrid.defaultProps = defaultProps;
PlaceholderGrid.propTypes = {
  className: _propTypes.default.string,
  classPrefix: _propTypes.default.string,
  rows: _propTypes.default.number,
  columns: _propTypes.default.number,
  rowHeight: _propTypes.default.number,
  rowMargin: _propTypes.default.number,
  active: _propTypes.default.bool
};
var _default = PlaceholderGrid;
exports.default = _default;
