// rsuite/cjs/utils/createComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _kebabCase = _interopRequireDefault(require("lodash/kebabCase"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

/**
 * Create a component with `classPrefix` and `as` attributes.
 * @param defaultProps
 */
function createComponent(defaultProps) {
  var name = defaultProps.name,
      componentAs = defaultProps.componentAs,
      componentClassPrefix = defaultProps.componentClassPrefix,
      componentRestProps = (0, _objectWithoutPropertiesLoose2.default)(defaultProps, ["name", "componentAs", "componentClassPrefix"]);

  var Component = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
    var Component = props.as,
        classPrefix = props.classPrefix,
        className = props.className,
        role = props.role,
        rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["as", "classPrefix", "className", "role"]);

    var _useClassNames = (0, _utils.useClassNames)(classPrefix),
        withClassPrefix = _useClassNames.withClassPrefix,
        merge = _useClassNames.merge;

    var classes = merge(className, withClassPrefix());
    return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({}, rest, {
      role: role,
      ref: ref,
      className: classes
    }));
  });

  Component.displayName = name;
  Component.propTypes = {
    as: _propTypes.default.elementType,
    className: _propTypes.default.string,
    classPrefix: _propTypes.default.string,
    children: _propTypes.default.node
  };
  Component.defaultProps = (0, _extends2.default)({}, componentRestProps, {
    as: componentAs || 'div',
    classPrefix: componentClassPrefix || (0, _kebabCase.default)(name)
  });
  return Component;
}

var _default = createComponent;
exports.default = _default;
