// rsuite/esm/Sidebar/Sidebar.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React, { useContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useClassNames } from '../utils';
import { ContainerContext } from '../Container/Container';
var defaultProps = {
  as: 'aside',
  classPrefix: 'sidebar',
  width: 260
};
var Sidebar = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Component = props.as,
      classPrefix = props.classPrefix,
      className = props.className,
      collapsible = props.collapsible,
      width = props.width,
      style = props.style,
      rest = _objectWithoutPropertiesLoose(props, ["as", "classPrefix", "className", "collapsible", "width", "style"]);

  var _useClassNames = useClassNames(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      merge = _useClassNames.merge;

  var classes = merge(className, withClassPrefix({
    collapse: collapsible
  }));

  var _useContext = useContext(ContainerContext),
      setHasSidebar = _useContext.setHasSidebar;

  useEffect(function () {
    /** Notify the Container that the Sidebar is in the child node of the Container. */
    setHasSidebar === null || setHasSidebar === void 0 ? void 0 : setHasSidebar(true);
  }, [setHasSidebar]);

  var styles = _extends({
    flex: "0 0 " + width + "px",
    width: width
  }, style);

  return /*#__PURE__*/React.createElement(Component, _extends({}, rest, {
    ref: ref,
    className: classes,
    style: styles
  }));
});
Sidebar.displayName = 'Sidebar';
Sidebar.defaultProps = defaultProps;
Sidebar.propTypes = {
  className: PropTypes.string,
  classPrefix: PropTypes.string,
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  collapsible: PropTypes.bool,
  style: PropTypes.object
};
export default Sidebar;
