// rsuite/esm/ButtonGroup/ButtonGroup.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import React from 'react';
import PropTypes from 'prop-types';
import { useClassNames } from '../utils';
var defaultProps = {
  as: 'div',
  classPrefix: 'btn-group',
  role: 'group'
};
var ButtonGroup = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var Component = props.as,
      classPrefix = props.classPrefix,
      role = props.role,
      className = props.className,
      children = props.children,
      block = props.block,
      vertical = props.vertical,
      justified = props.justified,
      size = props.size,
      rest = _objectWithoutPropertiesLoose(props, ["as", "classPrefix", "role", "className", "children", "block", "vertical", "justified", "size"]);

  var _useClassNames = useClassNames(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      merge = _useClassNames.merge;

  var classes = merge(className, withClassPrefix(size, {
    block: block,
    vertical: vertical,
    justified: justified
  }));
  return /*#__PURE__*/React.createElement(Component, _extends({}, rest, {
    role: role,
    ref: ref,
    className: classes
  }), children);
});
ButtonGroup.displayName = 'ButtonGroup';
ButtonGroup.defaultProps = defaultProps;
ButtonGroup.propTypes = {
  className: PropTypes.string,
  as: PropTypes.elementType,
  classPrefix: PropTypes.string,
  children: PropTypes.node,
  block: PropTypes.bool,
  vertical: PropTypes.bool,
  justified: PropTypes.bool,
  role: PropTypes.string,
  size: PropTypes.oneOf(['lg', 'md', 'sm', 'xs'])
};
export default ButtonGroup;
