// rsuite/cjs/List/ListItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

var _ListContext = _interopRequireDefault(require("./ListContext"));

var defaultProps = {
  as: 'div',
  classPrefix: 'list-item',
  collection: 0
};

var ListItem = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var Component = props.as,
      children = props.children,
      className = props.className,
      classPrefix = props.classPrefix,
      collection = props.collection,
      disabled = props.disabled,
      index = props.index,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["as", "children", "className", "classPrefix", "collection", "disabled", "index"]);

  var _useContext = (0, _react.useContext)(_ListContext.default),
      bordered = _useContext.bordered,
      register = _useContext.register,
      size = _useContext.size;

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      merge = _useClassNames.merge;

  var listItemRef = (0, _react.useRef)(null);
  (0, _react.useEffect)(function () {
    var _register = register({
      node: listItemRef.current,
      edgeOffset: null,
      info: {
        collection: collection,
        disabled: disabled,
        index: index
      }
    }),
        unregister = _register.unregister;

    return unregister;
  }, [collection, disabled, index, register]);
  var classes = merge(className, withClassPrefix(size, {
    disabled: disabled,
    bordered: bordered
  }));
  return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({
    role: "listitem",
    "aria-disabled": disabled
  }, rest, {
    ref: (0, _utils.mergeRefs)(listItemRef, ref),
    className: classes
  }), children);
});

ListItem.displayName = 'ListItem';
ListItem.defaultProps = defaultProps;
ListItem.propTypes = {
  index: _propTypes.default.number,
  collection: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  disabled: _propTypes.default.bool,
  children: _propTypes.default.node
};
var _default = ListItem;
exports.default = _default;
