// fixed-size-list/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mitt_1 = __importDefault(require("mitt"));
exports.eventCreated = 'FixedSizeListCreated';
exports.eventReset = 'FixedSizeListReset';
exports.eventNewItem = 'FixedSizeListNewItem';
exports.eventTruncate = 'FixedSizeListTruncate';
class FixedSizeList {
    constructor(maxSize, _list = [], eventEmitter = mitt_1.default()) {
        this.maxSize = maxSize;
        this._list = _list;
        this.eventEmitter = eventEmitter;
        this.eventEmitter.emit(exports.eventCreated, _list);
        this._truncate();
    }
    add(newEl) {
        this._list = [newEl, ...this._list];
        this._truncate();
        this.eventEmitter.emit(exports.eventNewItem, newEl);
        return this._list.length;
    }
    reset() {
        this._list = [];
        this.eventEmitter.emit(exports.eventReset);
    }
    on(event, cb) {
        this.eventEmitter.on(event, cb);
        return () => this.eventEmitter.off(event, cb);
    }
    get data() {
        return this._list;
    }
    get length() {
        return this._list.length;
    }
    _truncate() {
        if (this._list.length > this.maxSize) {
            const removedEls = this._list.splice(this.maxSize);
            this.eventEmitter.emit(exports.eventTruncate, removedEls);
        }
    }
}
exports.FixedSizeList = FixedSizeList;
